apiserver/admin.go

```diff
@@ -55,33 +86,88 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
 		return fail, errAlreadyLoggedIn
 	}
 
-	// apiRoot is the API root exposed to the client after authentication.
-	var apiRoot rpc.Root = newAPIRoot(
-		a.root.state,
-		a.srv.statePool,
-		a.srv.facades,
-		a.root.resources,
-		a.root,
-	)
+	// apiRoot is the API root exposed to the client after login.
+	apiRoot, err := rpcRoot(a.srv, a.root, req.AuthTag)
+	if err != nil {
+		return fail, errors.Trace(err)
+	}
 
-	// Use the login validation function, if one was specified.
-	if a.srv.validator != nil {
-		err := a.srv.validator(req)
-		switch err {
-		case params.UpgradeInProgressError:
-			apiRoot = restrictRoot(apiRoot, upgradeMethodsOnly)
-		case AboutToRestoreError:
-			apiRoot = restrictRoot(apiRoot, aboutToRestoreMethodsOnly)
-		case RestoreInProgressError:
-			apiRoot = restrictAll(apiRoot, restoreInProgressError)
-		case nil:
-			// in this case no need to wrap authed api so we do nothing
-		default:
-			return fail, errors.Trace(err)
+	authResult, err := a.authenticate(req)
+	if err, ok := errors.Cause(err).(*common.DischargeRequiredError); ok {
+		loginResult := params.LoginResult{
+			DischargeRequired:       err.Macaroon,
+			DischargeRequiredReason: err.Error(),
 		}
+		logger.Infof("login failed with discharge-required error: %v", err)
+		return loginResult, nil
+	}
+	if err != nil {
+		return fail, errors.Trace(err)
 	}
 
-	isUser := true
+	// Fetch the API server addresses from state.
+	hostPorts, err := a.root.state.APIHostPorts()
+	if err != nil {
+		return fail, errors.Trace(err)
+	}
+
+	model, err := a.root.state.Model()
+	if err != nil {
+		return fail, errors.Trace(err)
+	}
+
+	if authResult.userLogin {
+		switch model.MigrationMode() {
+		case state.MigrationModeImporting:
+			// The user is not able to access a model that is currently being
+			// imported until the model has been activated.
+			apiRoot = restrictAll(apiRoot, errors.New("migration in progress, model is importing"))
+		case state.MigrationModeExporting:
+			// The user is not allowed to change anything in a model that is
+			// currently being moved to another controller.
+			apiRoot = restrictRoot(apiRoot, migrationClientMethodsOnly)
+		}
+	}
+
+	loginResult := params.LoginResult{
+		Servers:       params.FromNetworkHostsPorts(hostPorts),
+		ControllerTag: model.ControllerTag().String(),
+		UserInfo:      authResult.userInfo,
+		ServerVersion: jujuversion.Current.String(),
+		PublicDNSName: a.srv.publicDNSName(),
+	}
+
+	if authResult.anonymousLogin {
```

Comment:
I'm not sure about the interaction of anonymous & controller. It should be possible to do an anonymous and contorller-only login I think? Isn't that the case for "crossmodelrelations"?

I think we want to combine the IsAnonymousFacade and IsControllerFacade filter predicates if both conditions are true?